### PR TITLE
Add per-mod settings

### LINKED ISSUE #6272: Allow mods to have their own settings
**Describe the new feature:** Allow mods to be customized with their own settings. For instance: https://github.com/ppy/osu/issues/4391, https://github.com/ppy/osu/issues/4175 and https://github.com/ppy/osu/issues/3201

Creating this issue for visibility as well as to discuss the implementation details.

**Proposal designs of the feature:** Create a new interface `IModHasSettings` which is different from `IReadFromConfig` as that persists through sessions and also an added benefit to avoid the need of creating new options and ending up cluttering the settings sidebar.

Mod settings will be found a sidebar alongside Input Settings and Visual Settings before starting the map. It's contents will be created from `IModHasSettings` method that should return a new drawable (or better yet implement a new class like `PlayerSettingItem` for streamlining) and the values of those controls will be read by the mod.

Mod settings will by dynamically created when we reach `PlayerLoader` depending on the mods selected

Current Implementation (Rate Mod as example)
image

### PR DESCRIPTION
**Pull request by LeNitrous**
Resolves #6272 and in turn allows #4391, #4175, and #3201 to be realized.

Creates a new interface `IModHasSettings` which allows a mod to create controls such as check boxes and sliders to allow the user control over it's properties.

A new button is found on `ModSelectOverlay` called "Customization" which opens a simple interface where all controls for customizing mods are found.

The Modify mod will be made in a separate PR and is only shown here as an example.

Example:
image

- [x] Depends on https://github.com/ppy/osu/pull/7081.

### CODE REVIEW

**Comment by iiSaLMaN** [comment]
Can the customization area not hit the bottom area of the mod select overlay (where the mod buttons are)

File: osu.Game/Rulesets/Mods/IModHasSettings.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+using osu.Framework.Graphics;
+
+namespace osu.Game.Rulesets.Mods
+{
+    /// <summary>
+    /// An interface for mods that allows user control over it's properties.
+    /// </summary>
+    public interface IModHasSettings
+    {
+        Drawable[] CreateControls();
```
Comment (iiSaLMaN) [inline osu.Game/Rulesets/Mods/IModHasSettings.cs]: This shouldn't pass a list of drawables, but rather, a list of settings controls maybe?

File: osu.Game/Rulesets/Mods/Mod.cs [hunk 2]
```diff
@@ -69,7 +69,7 @@ public abstract class Mod : IMod, IJsonSerializable
         /// <summary>
         /// Creates a copy of this <see cref="Mod"/> initialised to a default state.
         /// </summary>
-        public virtual Mod CreateCopy() => (Mod)Activator.CreateInstance(GetType());
+        public virtual Mod CreateCopy() => (Mod)MemberwiseClone();
```
Comment (iiSaLMaN) [inline osu.Game/Rulesets/Mods/Mod.cs]: This should likely be in a different PR, kind-of out of scope for what this is implementing.

**Review by frenzibyte — CHANGES_REQUESTED** [review]

File: osu.Game/Rulesets/Mods/IModHasSettings.cs (hunk 1 above)
Comment (LeNitrous) [reply osu.Game/Rulesets/Mods/IModHasSettings.cs]: I'd like to automatically generate it's settings based on the class' properties (maybe marked with an attribute?) but I'm not sure of that approach just yet.

File: osu.Game/Rulesets/Mods/Mod.cs (hunk 2 above)
Comment (bdach) [reply osu.Game/Rulesets/Mods/Mod.cs]: I think it's fine here, without this any mods with internal state would be thoroughly broken.

Comment (iiSaLMaN) [reply osu.Game/Rulesets/Mods/Mod.cs]: This code is about fixing an issue when doing something to the mods as discussed, this should've been in another PR, just saying for the author to take note of that later on.

**Comment by peppy** [comment]
This is missing any form of testing. How are you testing? Please make a test scene with a test mod which implements settings.

**Comment by peppy** [comment]
Just a heads up that whatever you did to this PR to bring it up-to-date is not the correct way (`git merge upstream/master` is what you want in the future).

**Comment by peppy** [comment]
Updated behaviour somewhat to use new logic in #7081.

File: osu.Game/Configuration/SettingSourceAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+    public static class SettingSourceExtensions
+    {
+        public static IEnumerable<Drawable> CreateSettingsControls(this object obj)
+        {
+            var configProperties = obj.GetType().GetProperties(BindingFlags.GetProperty | BindingFlags.Public | BindingFlags.Instance).Where(p => p.GetCustomAttribute<SettingSourceAttribute>(true) != null);
+
+            foreach (var property in configProperties)
+            {
+                var attr = property.GetCustomAttribute<SettingSourceAttribute>(true);
```
Comment (huoyaoyuan) [inline osu.Game/Configuration/SettingSourceAttribute.cs]: Logic neat:
```csharp
foreach (var property in obj.GetType.GetProperties(...))
{
    var attr = property.GetCustomAttribute<SettingSourceAttribute>(true);
    if (attr == null) continue;


}
```

File: osu.Game/Configuration/SettingSourceAttribute.cs [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+                        yield return new SettingsTextBox
+                        {
+                            LabelText = attr.Label,
+                            Bindable = bString
+                        };
+
+                        break;
+
+                    case IBindable bindable:
+
```
Comment (huoyaoyuan) [inline osu.Game/Configuration/SettingSourceAttribute.cs]: Too many blank lines here.

File: osu.Game.Tests/Visual/UserInterface/TestSceneModSettings.cs [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+                    if (section.ModType == ModType.Conversion)
+                    {
+                        section.Mods = new Mod[]
+                        {
+                            new TestModCustomisable1(),
+                            new TestModCustomisable2()
+                        };
+                    }
+                    else
+                        section.Mods = new Mod[] { };
```
Comment (huoyaoyuan) [inline osu.Game.Tests/Visual/UserInterface/TestSceneModSettings.cs]: `Array.Empty<>`. Can be addressed later.

File: osu.Game/Configuration/SettingSourceAttribute.cs (hunk 3 above)
Comment (peppy) [reply osu.Game/Configuration/SettingSourceAttribute.cs]: Applied in https://github.com/ppy/osu/pull/7081

File: osu.Game/Configuration/SettingSourceAttribute.cs (hunk 4 above)
Comment (peppy) [reply osu.Game/Configuration/SettingSourceAttribute.cs]: Applied in #7081

File: osu.Game/Rulesets/Mods/Mod.cs (hunk 2 above)
Comment (smoogipoo) [inline osu.Game/Rulesets/Mods/Mod.cs]: I'm still not quite sure about this. I believe in its current state it will break the easy mod.

What we really want here is to bind bindables between mods, so perhaps a custom implementation that _does_ use `CreateInstance` and also reflection to bind (or copy? unsure) bindables.

File: osu.Game/Overlays/Mods/ModSelectOverlay.cs [hunk 6]
```diff
@@ -284,10 +331,32 @@ private void load(OsuColour colours, IBindable<RulesetInfo> ruleset, AudioManage
             Ruleset.BindTo(ruleset);
             if (mods != null) SelectedMods.BindTo(mods);
 
+            SelectedMods.ValueChanged += updateModSettings;
+            Ruleset.ValueChanged += _ => ModSettingsContent.Clear();
```
Comment (smoogipoo) [inline osu.Game/Overlays/Mods/ModSelectOverlay.cs]: Is this required? It should be handled by the `SelectedMods` bindable?

File: osu.Game/Overlays/Mods/ModSelectOverlay.cs [hunk 7]
```diff
@@ -226,6 +232,17 @@ public ModSelectOverlay()
                                                     Right = 20
                                                 }
                                             },
+                                            CustomiseButton = new TriangleButton
+                                            {
+                                                Width = 180,
+                                                Text = "Customization",
```
Comment (smoogipoo) [inline osu.Game/Overlays/Mods/ModSelectOverlay.cs]:
```suggestion
                                                Text = "Customisation",
```

File: osu.Game/Overlays/Mods/ModSelectOverlay.cs [hunk 8]
```diff
@@ -284,10 +331,32 @@ private void load(OsuColour colours, IBindable<RulesetInfo> ruleset, AudioManage
             Ruleset.BindTo(ruleset);
             if (mods != null) SelectedMods.BindTo(mods);
 
+            SelectedMods.ValueChanged += updateModSettings;
```
Comment (smoogipoo) [inline osu.Game/Overlays/Mods/ModSelectOverlay.cs]: This should probably be a `BindValueChanged(, true)`.

Also this shouldn't be inside `load()`.

Also this should probably not even exist, and `updateModSettings()` should be invoked by `selectedModsChanged` (see the existing `LoadComplete()`).

**Review by smoogipoo — CHANGES_REQUESTED** [review]

File: osu.Game/Rulesets/Mods/Mod.cs (hunk 2 above)
Comment (smoogipoo) [reply osu.Game/Rulesets/Mods/Mod.cs]: Actually it probably won't break the easy mod since it's cloned with a zero-state, still feels kind of dodgy though.

**Review by smoogipoo — APPROVED** [review]
